In [1]:
import pandas as pd
import logging

# Log fayl manzili
log_path = r"C:\Users\Rasulbek907\Desktop\Project_MP\Log\data_loader.log"

# Log sozlamalari
logging.basicConfig(
    filename=log_path,
    filemode='a',  # Append mode
    format='%(asctime)s - %(levelname)s - %(message)s',
    level=logging.INFO
)

try:
    logging.info("CSV fayl o'qilmoqda:...")
    df = pd.read_csv(r"C:\Users\Rasulbek907\Desktop\Project_MP\Data\Feature_Selection\Filtered_Features.csv")
    logging.info(f"Fayl muvaffaqiyatli o'qildi. Satırlar soni: {len(df)} ustunlar soni: {len(df.columns)}")
except Exception as e:
    logging.error(f"CSV faylni o'qishda xatolik: {e}")

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unnamed: 0           4209 non-null   int64  
 1   id                   4209 non-null   float64
 2   ref                  4209 non-null   float64
 3   subtitle             4209 non-null   float64
 4   creatorname          4209 non-null   float64
 5   totalbytes           4209 non-null   float64
 6   lastupdated          4209 non-null   float64
 7   downloadcount        4209 non-null   float64
 8   title                4209 non-null   float64
 9   tags                 4209 non-null   float64
 10  cluster              4209 non-null   int64  
 11  anomaly              4209 non-null   float64
 12  download_view_ratio  4209 non-null   float64
 13  num_tags             4209 non-null   float64
 14  title_length         4209 non-null   float64
 15  subtitle_length      4209 non-null   f

In [3]:
df.head()

Unnamed: 0        id       ref  subtitle  creatorname  totalbytes  \
0           0  1.652531 -1.497901 -1.343173    -1.482264   -0.205871   
1           1  1.608563 -1.581849  0.951414    -1.578557   -0.205876   
2           2  1.672062 -1.224657 -0.147321    -1.218896   -0.205877   
3           3  1.690704 -1.434528 -1.140709    -1.420537   -0.205814   
4           4  1.583257 -1.599133 -1.353049    -1.588433   -0.205741   

   lastupdated  downloadcount     title      tags  ...   anomaly  \
0     1.440289      -0.334293 -1.148158 -1.305056  ...  0.333377   
1     1.295437       0.105377 -0.349943  0.271367  ...  0.333377   
2     1.522592      -0.276472  0.955542  1.099529  ...  0.333377   
3     1.599133      -0.329493 -1.527142 -0.562436  ...  0.333377   
4     1.204905       0.203582 -1.403888 -0.790862  ...  0.333377   

   download_view_ratio  num_tags  title_length  subtitle_length  \
0             0.054731  0.644314      0.521569         1.310725   
1             0.015678  0.321751      1.770817         0.682214   
2             0.080659  0.063700      0.789265         0.853626   
3             0.141278  0.708827      0.253873         0.796489   
4             0.161541  0.902365      0.878497         1.310725   

   creator_popularity  owner_popularity  lastupdated_month  lastupdated_day  \
0                 0.0          -0.06614           0.890462         0.152226   
1                 0.0          -0.06614           0.890462        -1.626624   
2                 0.0          -0.06614           0.890462         1.152829   
3                 0.0          -0.06614           1.189175        -1.515446   
4                 0.0          -0.06614           0.591749         0.485760   

   lastupdated_weekday  
0             0.535769  
1            -0.467145  
2             1.538684  
3             1.538684  
4             1.037227  

[5 rows x 21 columns]

# Bagging + Classification 

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import BaggingClassifier, RandomForestClassifier, ExtraTreesClassifier

In [5]:
x=df.drop('cluster',axis=1)
y=df['cluster'].astype(int)
X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state=42)

# Kop ishlatilanadigan xolatlar

In [6]:
# Random Forest
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)
print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))

Random Forest Accuracy: 0.9865399841646872


In [7]:
# 2. Bagged Decision Trees

bag_dt = BaggingClassifier(
    estimator=DecisionTreeClassifier(),
    n_estimators=100,
    random_state=42
)
bag_dt.fit(X_train, y_train)
y_pred_bagdt = bag_dt.predict(X_test)
print("Bagged Decision Trees Accuracy:", accuracy_score(y_test, y_pred_bagdt))

Bagged Decision Trees Accuracy: 0.9881235154394299


In [9]:
# 3. Extra Trees

extra_trees = ExtraTreesClassifier(n_estimators=100, random_state=42)
extra_trees.fit(X_train, y_train)
y_pred_extra = extra_trees.predict(X_test)
print("Extra Trees Accuracy:", accuracy_score(y_test, y_pred_extra))

Extra Trees Accuracy: 0.9786223277909739


# Kam ishlatilanadigan xolatlar

In [10]:
# 4. Bagging KNN

bag_knn = BaggingClassifier(
    estimator=KNeighborsClassifier(n_neighbors=3),
    n_estimators=50,
    random_state=42
)
bag_knn.fit(X_train, y_train)
y_pred_knn = bag_knn.predict(X_test)
print("Bagged KNN Accuracy:", accuracy_score(y_test, y_pred_knn))

Bagged KNN Accuracy: 0.9398258115597783


##  Bagging Classifier 
 
| Model | Score | Izoh |
|-------|-------|------|
| Bagging Classifier (Random Forest) | <span style="color:green">0.98</span> | Mukammal |
| Bagging Classifier (Bagged DT) | <span style="color:green">0.98</span> | Mukammal |
| Bagging Classifier (Extra Trees) | <span style="color:green">0.97</span> | Mukammal |
| Bagging Classifier (Bagged KNN) | <span style="color:red">0.93</span> | Mukammal |

